# Descriptive Analysis of User Data from Medical Image Labeling Mobile Application

## Analysis Objective: 
Complie a descriptive analysis supporting the robustness of the crowdsourced opinion. Justify downsizing the amount of experts needed to develop a labeled medical image database for artifical intelligance to train on. 
### Dataset Background

•User interaction data was extracted from mobile application where tasked to label whether a medical image appears normal or abnormal. 

•Two user categories were differentiated:
1. Crowd, anyone who downloaded and interacted with the application
2. Medical experts hired to interact with the application



# Data Prep and Cleaning

### Import datasets

In [1]:
import pandas as pd
import numpy as np
results = pd.read_csv('1345_customer_results.csv') #medical case results

### Inspect Customer Results

In [2]:
results.dtypes
results = results.set_index('Case ID')

In [3]:
results = results.dropna(subset=['Origin']) 
results["Expert: Abnormal Votes"] = results["Origin"].str.extract(r'vote(\d)').astype(float)
results = results.drop(['Origin Created At','Origin','Content ID','URL'],axis=1)

Identifying data filtered for security purposes

In [4]:
results.head(2)

Labeling State  Series  Series Index  Patch  Qualified Reads  \
Case ID                                                                
5888087  Gold Standard     NaN           NaN    NaN                2   
5888088  Gold Standard     NaN           NaN    NaN                3   

        Correct Label Majority Label  Difficulty  Agreement  \
Case ID                                                       
5888087          'no'           'no'         0.0        1.0   
5888088          'no'           'no'         0.0        1.0   

        First Choice Answer  First Choice Votes  First Choice Weight  \
Case ID                                                                
5888087                'no'                   2                 1.54   
5888088                'no'                   3                 2.34   

        Second Choice Answer  Second Choice Votes  Second Choice Weight  \
Case ID                                                                   
5888087                'yes'                    0                   0.0   
5888088                'yes'                    0                   0.0   

         Internal Notes Comments  Explanation  Expert: Abnormal Votes  
Case ID                                                                
5888087             NaN       []          NaN                     2.0  
5888088             NaN       []          NaN                     0.0

Any rows that did not have a string associated with expert votes in the URL were dropped (i.e. NA)

In [5]:
results = results.dropna(subset=["Expert: Abnormal Votes"])

In [6]:
results.loc[results['Series'].notna()| results['Series Index'].notna() | results['Patch'].notna() | results['Internal Notes'].notna() | results['Explanation'].notna()]

Empty DataFrame
Columns: [Labeling State, Series, Series Index, Patch, Qualified Reads, Correct Label, Majority Label, Difficulty, Agreement, First Choice Answer, First Choice Votes, First Choice Weight, Second Choice Answer, Second Choice Votes, Second Choice Weight, Internal Notes, Comments, Explanation, Expert: Abnormal Votes]
Index: []

**Inspect NaN Columns for Content**: Dataframe is empty; columns inspected will be dropped


In [7]:
results = results.drop(['Series','Series Index','Patch','Internal Notes','Explanation'],axis=1)

In [8]:
results[results['Comments'] != '[]']


Labeling State  Qualified Reads Correct Label Majority Label  \
Case ID                                                                
5892332  Gold Standard                1          'no'           'no'   
5894116  Gold Standard                5          'no'          'yes'   
5896433  Gold Standard                3         'yes'           'no'   
5899520  Gold Standard                2         'yes'           'no'   
5900998  Gold Standard                2          'no'          'yes'   
5901914  Gold Standard                6         'yes'           'no'   
5902040  Gold Standard                2          'no'          'yes'   
5904120  Gold Standard                1         'yes'           'no'   
5904413  Gold Standard                3         'yes'           'no'   
5904472  Gold Standard                3         'yes'          'yes'   
5905734  Gold Standard                4         'yes'          'yes'   
5908501  Gold Standard                5          'no'          'yes'   
5909550  Gold Standard                3         'yes'          'yes'   
5911442  Gold Standard                2          'no'           'no'   
5913944  Gold Standard                3          'no'           'no'   
5916678  Gold Standard                2         'yes'          'yes'   
5916911  Gold Standard                4         'yes'           'no'   
5916943  Gold Standard                4          'no'           'no'   
5917576  Gold Standard                3          'no'          'yes'   

         Difficulty  Agreement First Choice Answer  First Choice Votes  \
Case ID                                                                  
5892332       0.000      1.000                'no'                   1   
5894116       1.000      1.000               'yes'                   5   
5896433       1.000      1.000                'no'                   3   
5899520       1.000      1.000                'no'                   2   
5900998       1.000      1.000               'yes'                   2   
5901914       1.000      1.000                'no'                   6   
5902040       1.000      1.000               'yes'                   2   
5904120       1.000      1.000                'no'                   1   
5904413       1.000      1.000                'no'                   3   
5904472       0.333      0.667               'yes'                   2   
5905734       0.500      0.500               'yes'                   2   
5908501       1.000      1.000               'yes'                   5   
5909550       0.000      1.000               'yes'                   3   
5911442       0.500      0.500               'yes'                   1   
5913944       0.333      0.667                'no'                   2   
5916678       0.000      1.000               'yes'                   2   
5916911       1.000      1.000                'no'                   4   
5916943       0.250      0.750                'no'                   3   
5917576       1.000      1.000               'yes'                   3   

         First Choice Weight Second Choice Answer  Second Choice Votes  \
Case ID                                                                  
5892332                 0.80                'yes'                    0   
5894116                 4.00                 'no'                    0   
5896433                 2.32                'yes'                    0   
5899520                 1.58                'yes'                    0   
5900998                 1.56                 'no'                    0   
5901914                 4.72                'yes'                    0   
5902040                 1.58                 'no'                    0   
5904120                 0.78                'yes'                    0   
5904413                 2.46                'yes'                    0   
5904472                 1.58                 'no'                    1   
5905734                 1.54                 'no'   

**Inspect Comments for Relevance**: None of the comments seem relevant; comments column will be dropped

In [9]:
results = results.drop(['Comments'],axis=1)

**There should only be 8 experts total**: drop cases for expert count greater than 8

In [10]:
results = results[results["Expert: Abnormal Votes"] <= 8]

In [89]:
results = results.drop(['First Choice Answer','First Choice Votes','First Choice Weight','Second Choice Answer','Second Choice Votes','Second Choice Weight'], axis=1)

**Drop other columns not used for descriptive analysis**: Nth Choice Answer; Nth Choice Votes; Nth Choice Weight

### Important columns for analysis
Each row corresponds to a medical case 

**Identifiers:** 

Case ID: unique identifier will serve as index

Labeling State: identifies whether a expert consensus has been achieved (yes=Gold Standard, no= In Progress)

URL: Extracted out expert vote count within the URL 

**Reads and Annotations**

Qualified Reads: total crowd vote count

Expert: Abnormal Votes: number of experts who thought the case was abnormal

(note, the total of experts voting is always 8)

Correct Label: overall expert consensus 

{yes=case is abnormal, no=case is normal, NaN=no consensus}

Majority Label: overall crowd consensus on each case

**Measures of Confidence**

Difficulty: Qualified Reads *without the Correct Label* divided by total Qualified Reads.

Agreement: Qualified Reads *with the Majority Label* divided by total Qualified Reads.

        
        



### Add Additional Relevant Columns and Optimize Dataframe




In [12]:
bins=[0,0.2,0.4,0.6,0.8,1]
labels=['very easy','easy','moderate','challenging','very challenging']
results['Difficulty Category'] = pd.cut(results['Difficulty'],bins=bins,labels=labels,include_lowest=True)

#### Difficulty Category: 
Clustered cases categorically based on difficulty scores calculated from crowd performance

#### Expert: Normal Votes:
Subtracted the number of total experts by the known number of experts who voted the case as abnormal

#### Expert Agreement: 
Divided the number of experts who voted the case as abnormal by the total number of experts; result is the porportion of experts who agree that the case is abnormal.

#### Error Rate: 
Extracted the indexes for each category and calculated the "error rate" for the experts; result is the porportion of experts who did not vote for the expert majority

#### Consensus:
Indiciated cases where there was unanimity between experts and crowd.

### I will rename some of the original columns for clarity

   #### {Original column --> Renamed Column}
    
    Correct Label --> Expert Majority

    Majority Label --> Crowd Majority

    Difficulty --> Expert/Crowd Disagreement

    Agreement --> Crowd Agreement

In [13]:
df = results
df["Expert Majority"] = results["Correct Label"]
df["Crowd Majority"] = results["Majority Label"]
df["Expert/Crowd Disagreement"] = results["Difficulty"] 
df["Crowd Agreement"] = results["Agreement"] 
df = df.drop(columns= ["Correct Label","Majority Label","Difficulty","Agreement"])

In [14]:
expert_count = 8
df["Expert: Normal Votes"] = (expert_count - results["Expert: Abnormal Votes"])
df["Expert Agreement"] = df["Expert: Abnormal Votes"]/expert_count
df['Consensus'] = np.where(df['Expert Majority'] == df['Crowd Majority'],'yes','no')

In [15]:
EM_yes = df.index[df['Expert Majority'] == "'yes'"].tolist()

EM_no = df.index[df['Expert Majority'] == "'no'"].tolist()

df.loc[EM_yes,"Error Rate"]= df['Expert: Normal Votes'][EM_yes]/expert_count
df.loc[EM_no,"Error Rate"]= df['Expert: Abnormal Votes'][EM_no]/expert_count


In [91]:
#df.plot(kind= 'scatter',x = 'Qualified Reads', y ='Expert/Crowd Disagreement')
#filt_df.plot(kind= 'scatter',x = 'Qualified Reads', y ='Expert/Crowd Disagreement')
import plotly.express as px
#fig = px.histogram(df,x= "Qualified Reads", color="Error Rate")

fig = px.density_heatmap(df, x="Qualified Reads", y='Expert/Crowd Disagreement', title='Group Size is Positively Associated with Correct Labeling')
fig.update_layout(xaxis_range=[0.5,11])
fig.update_layout(yaxis_range=[0.45,1.05])
fig.update_layout(yaxis_title='Answered Incorrectly')
fig.update_layout(xaxis_title='Number of Crowd Users')
fig.show()


When group size is larger, seen with the uniform dark blue color transition, the frequency of half the group or more answering incorrectly is nearly zero. **The number of qualified reads will be truncated to 5 or more per case to improve crowd accuracy.**

In [80]:
filt_df = df[df['Qualified Reads'] >= 5]
filt_df

Labeling State  Qualified Reads First Choice Answer  \
Case ID                                                       
5888091    In Progress                7               'yes'   
5888093  Gold Standard                6                'no'   
5888099  Gold Standard                5                'no'   
5888100  Gold Standard                7                'no'   
5888102  Gold Standard                6                'no'   
...                ...              ...                 ...   
5918367  Gold Standard                5               'yes'   
5918368  Gold Standard                8               'yes'   
5918372  Gold Standard                7               'yes'   
5918377    In Progress                6               'yes'   
5918379  Gold Standard                5               'yes'   

         First Choice Votes  First Choice Weight Second Choice Answer  \
Case ID                                                                 
5888091                   4                 3.28                 'no'   
5888093                   6                 4.94                'yes'   
5888099                   5                 4.05                'yes'   
5888100                   6                 4.92                'yes'   
5888102                   6                 4.66                'yes'   
...                     ...                  ...                  ...   
5918367                   5                 3.98                 'no'   
5918368                   8                 6.50                 'no'   
5918372                   7                 5.78                 'no'   
5918377                   6                 4.78                 'no'   
5918379                   4                 3.20                 'no'   

         Second Choice Votes  Second Choice Weight  Expert: Abnormal Votes  \
Case ID                                                                      
5888091                    3                  2.32                     4.0   
5888093                    0                  0.00                     0.0   
5888099                    0                  0.00                     0.0   
5888100                    1                  0.80                     0.0   
5888102                    0                  0.00                     0.0   
...                      ...                   ...                     ...   
5918367                    0                  0.00                     3.0   
5918368                    0                  0.00                     5.0   
5918372                    0                  0.00                     6.0   
5918377                    0                  0.00                     4.0   
5918379                    1                  0.76                     5.0   

        Difficulty Category Expert Majority Crowd Majority  \
Case ID                                                      
5888091                 NaN             NaN          'yes'   
5888093           very easy            'no'           'no'   
5888099           very easy            'no'           'no'   
5888100           very easy            'no'           'no'   
5888102           very easy            'no'           'no'   
...                     ...             ...            ...   
5918367    very challenging            'no'          'yes'   
5918368           very easy           'yes'          'yes'   
5918372           very easy           'yes'          'yes'   
5918377                 NaN             NaN          'yes'   
5918379           very easy           'yes'          'yes'   

         Expert/Crowd Disagreement  Crowd Agreement  Expert: Normal Votes  \
Case ID                                                                     
5888091                        NaN            0.571                   4.0   
5888093                      0.000            1.000                   8.0   
5888099                      0.000            1.000                   8.0   
5888100                      0.

# Exploratory Analysis 

In [85]:
#%pip install jupyter-dash
import plotly.io as pio
import plotly.figure_factory as ff
pio.renderers.default='notebook'
import matplotlib.pyplot as plt
import plotly.graph_objects as go

In [86]:

index = filt_df.index[filt_df['Crowd Majority']=="'no'"].tolist()
filt_df['Crowd Agreement'][index] -= 0.5
filt_df

Labeling State  Qualified Reads First Choice Answer  \
Case ID                                                       
5888091    In Progress                7               'yes'   
5888093  Gold Standard                6                'no'   
5888099  Gold Standard                5                'no'   
5888100  Gold Standard                7                'no'   
5888102  Gold Standard                6                'no'   
...                ...              ...                 ...   
5918367  Gold Standard                5               'yes'   
5918368  Gold Standard                8               'yes'   
5918372  Gold Standard                7               'yes'   
5918377    In Progress                6               'yes'   
5918379  Gold Standard                5               'yes'   

         First Choice Votes  First Choice Weight Second Choice Answer  \
Case ID                                                                 
5888091                   4                 3.28                 'no'   
5888093                   6                 4.94                'yes'   
5888099                   5                 4.05                'yes'   
5888100                   6                 4.92                'yes'   
5888102                   6                 4.66                'yes'   
...                     ...                  ...                  ...   
5918367                   5                 3.98                 'no'   
5918368                   8                 6.50                 'no'   
5918372                   7                 5.78                 'no'   
5918377                   6                 4.78                 'no'   
5918379                   4                 3.20                 'no'   

         Second Choice Votes  Second Choice Weight  Expert: Abnormal Votes  \
Case ID                                                                      
5888091                    3                  2.32                     4.0   
5888093                    0                  0.00                     0.0   
5888099                    0                  0.00                     0.0   
5888100                    1                  0.80                     0.0   
5888102                    0                  0.00                     0.0   
...                      ...                   ...                     ...   
5918367                    0                  0.00                     3.0   
5918368                    0                  0.00                     5.0   
5918372                    0                  0.00                     6.0   
5918377                    0                  0.00                     4.0   
5918379                    1                  0.76                     5.0   

        Difficulty Category Expert Majority Crowd Majority  \
Case ID                                                      
5888091                 NaN             NaN          'yes'   
5888093           very easy            'no'           'no'   
5888099           very easy            'no'           'no'   
5888100           very easy            'no'           'no'   
5888102           very easy            'no'           'no'   
...                     ...             ...            ...   
5918367    very challenging            'no'          'yes'   
5918368           very easy           'yes'          'yes'   
5918372           very easy           'yes'          'yes'   
5918377                 NaN             NaN          'yes'   
5918379           very easy           'yes'          'yes'   

         Expert/Crowd Disagreement  Crowd Agreement  Expert: Normal Votes  \
Case ID                                                                     
5888091                        NaN            0.571                   4.0   
5888093                      0.000            0.500                   8.0   
5888099                      0.000            0.500                   8.0   
5888100                      0.

0.5 was subtracted to match how expert agreement was calculated.

## When experts and crowd disagree, what's the extent of their division?

In [140]:
consensus_no = filt_df.index[filt_df['Consensus']=='no'].tolist()
split_conses = (df[df["Expert Agreement"]==0.5])


fig2 = px.violin(filt_df.loc[consensus_no], 
                    x='Expert Agreement',color='Expert Majority',
                    color_discrete_map={"'yes'":'purple',"'no'":'red'}, 
                    labels={'x' : 'Agreement Ratio', 'y' : 'Count'}
                   )
fig1 = px.violin(filt_df.loc[consensus_no], 
                    x='Crowd Agreement',color='Crowd Majority',
                    color_discrete_map={"'yes'":'green',"'no'":'yellow'})

fig2.update_xaxes(dtick=0.2)
fig2.update_traces(opacity=0.52)
fig1.update_traces(opacity=0.52)
#fig2.update_layout(title_text='Overall Vote Distribution',
   # title_x=0.5, showlegend=True,
   # legend_title=None)
fig1.data[0].name="Crowd Majority: Yes"
fig2.data[0].name="Expert Majority: No"
fig1.data[1].name="Crowd Majority: No"
fig2.data[1].name="Expert Majority: Yes"
                  
fig2.add_trace(fig1.data[0])
fig2.add_trace(fig1.data[1])

fig2.update_layout(title_text='Normal Votes Distribution',
    showlegend=True,
    legend_title=None)
fig2.update_xaxes(range=[-0.02,0.42])
fig2.update_layout(xaxis_title='Agreement')
fig2.show()
fig2_copy = fig2
fig2_copy.update_xaxes(range=[0.58,1.02])

fig2.update_layout(title_text='Abnormal Votes Distribution',
    showlegend=True,
    legend_title=None)
fig2_copy.show()





print((743+116)/len(filt_df.index)*100)#crowd
print((48+89)/len(filt_df.index)*100) #experts


5.770135017129039
0.9202660038960166


•Axis range represents the extent of agreement within a group: 0.5 is indecisively split; 1 is full group agreement that the case is normal; 0 is full group agreement that the case is abnormal.

•Density of occurrences is large for the crowd group and very narrow for experts for unanimous agreement that the case is abnormal (when Agreement=1).The difference in density is most significant here.

•Division between experts for determining if the case is frequently split for cases that are either normal or abnormal, more so than crowd. 

•743 cases were confidently labeled by crowd as abnormal (=1) and 116 as normal (=0); 48 cases were confidently labeled by experts as abnormal and 89 cases as normal.

•5.77% of total cases were able to be labeled by the crowd with full confidence. Only 0.92% of total cases were able to be labeled by experts with full confidence


•These observations point to how the crowd agrees more frequently with the overall expert conclusions for each case more so than individual experts agreeing with the expert majority, supporting the claim that crowd aggregate can more accurately choose than individual experts can. **Therefore, the amount of experts needed to train artifical intelligance can be fewer, allowing the company to save costs and resources without compromising accuracy.**